### ISSUE TITLE
Split BaseEngine (or make it more granular wrt to operations)

### ISSUE THREAD

**Issue by runcom** [ansible/ansible-container#330]
##### ISSUE TYPE

 - Feature Idea

Right now, a `BasicEngine` is in charge of building, running/stopping/restarting/etc, pushing and maybe other stuff.
This is crystal clear when you think of a full blown engine like docker. However, there are tools out there which are able to just do a subset of what a whole docker engine can do.
Examples:
- push/pull ([skopeo](https://github.com/projectatomic/skopeo) 
- build ([box](https://github.com/erikh/box) and others probably)
- run/stop/restart/etc ([runc](https://github.com/opencontainers/runc)
- more?

Would it be possible to split the current implementation or provide a more granular way to splitting operations?

**Comment by j00bar** [comment]
In His wisdom, Guido gave us mixins. That would be my suggestion for this.

**Comment by runcom** [comment]
> In His wisdom, Guido gave us mixins. That would be my suggestion for this.

I understand that, yet `BaseEngine` seems to assume the base engine is docker? i.e. https://github.com/ansible/ansible-container/blob/develop/container/engine.py#L233 this method is really just for docker for what I can see (and know). Docker stores login info on disk at `$HOME/.docker/config.json` but usually you don't really need this at all (you just login and send auth tokens to a registry)

**Comment by gregdek** [comment]
I know @j00bar has some ideas about this, maybe worth discussion sometime in the 0.4 timeframe. One key question is: @runcom are you actually looking at writing a new BasicEngine for one of these targets, and if so, on what time line? That will help us figure out when we should attack this -- i.e. 0.4, or later?

**Comment by runcom** [comment]
@gregdek yes, we have plans to add a similar feature for other engines (skopeo for pulling etc etc). I'm not really sure on the time-frame from this, @rhatdan do you?

**Comment by rhatdan** [comment]
@gregdek We are just at the investigation stage at this point.  I would figure over the next couple of months, we  could free up some people to work on it.

**Comment by gregdek** [comment]
May be worth convening a requirements meeting to discuss, if we can get some hands to work on it. Targeting mid-February very tentatively for 0.4 at this point, so maybe worth a discussion in January sometime -- but best to leave that up to @j00bar I think.

**Comment by vbatts** [comment]
@chouseknecht This is what you and I were talking about last week
